File: code/Back/schemas/solicitacao.py
from datetime import date


STATUS_VALIDOS = [
    "PENDENTE",
    "ACEITA",
    "EM_ANDAMENTO",
    "CONCLUIDA",
    "CANCELADA",
    "RECUSADA"
]


def validar_solicitacao_create(data):
    campos_obrigatorios = [
        "cliente_id",
        "titulo",
        "descricao",
        "tipo_servico"
    ]

    for campo in campos_obrigatorios:
        if campo not in data or data[campo] in [None, ""]:
            return False, f"O campo '{campo}' é obrigatório."

    # Validações de tipo
    try:
        int(data["cliente_id"])
    except (ValueError, TypeError):
        return False, "cliente_id deve ser um inteiro."

    if "orcamento" in data and data["orcamento"] is not None:
        try:
            float(data["orcamento"])
        except (ValueError, TypeError):
            return False, "orcamento deve ser um número."

    if "prazo" in data and data["prazo"]:
        try:
            date.fromisoformat(data["prazo"])
        except (ValueError, TypeError):
            return False, "prazo deve estar no formato YYYY-MM-DD."

    return True, None


def validar_status(status):
    if status not in STATUS_VALIDOS:
        return False, f"Status inválido. Use um destes: {', '.join(STATUS_VALIDOS)}"

    return True, None


def solicitacao_to_dict(solicitacao):
    return {
        "id": solicitacao.id,
        "cliente_id": solicitacao.cliente_id,
        "prestador_id": solicitacao.prestador_id,
        "titulo": solicitacao.titulo,
        "descricao": solicitacao.descricao,
        "tipo_servico": solicitacao.tipo_servico,
        "orcamento": solicitacao.orcamento,
        "prazo": solicitacao.prazo.isoformat() if solicitacao.prazo else None,
        "referencia": solicitacao.referencia,
        "status": solicitacao.status,
        "criado_em": solicitacao.criado_em.isoformat() if solicitacao.criado_em else None,
        "atualizado_em": solicitacao.atualizado_em.isoformat() if solicitacao.atualizado_em else None
    }

File: code/Mobile/prestador/lib/models/solicitacao.dart
class Solicitacao {
  const Solicitacao({
    this.id,
    required this.clienteId,
    this.prestadorId,
    required this.titulo,
    required this.descricao,
    required this.tipoServico,
    this.orcamento,
    this.prazo,
    this.referencia,
    required this.status,
    this.criadoEm,
    this.atualizadoEm,
  });

  final int? id;
  final int clienteId;
  final int? prestadorId;
  final String titulo;
  final String descricao;
  final String tipoServico;
  final double? orcamento;
  final String? prazo;
  final String? referencia;
  final String status;
  final String? criadoEm;
  final String? atualizadoEm;

  factory Solicitacao.fromJson(Map<String, dynamic> json) {
    return Solicitacao(
      id: json['id'] as int?,
      clienteId: json['cliente_id'] as int,
      prestadorId: json['prestador_id'] as int?,
      titulo: json['titulo'] as String,
      descricao: json['descricao'] as String,
      tipoServico: json['tipo_servico'] as String,
      orcamento: (json['orcamento'] as num?)?.toDouble(),
      prazo: json['prazo'] as String?,
      referencia: json['referencia'] as String?,
      status: json['status'] as String,
      criadoEm: json['criado_em'] as String?,
      atualizadoEm: json['atualizado_em'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'cliente_id': clienteId,
      'prestador_id': prestadorId,
      'titulo': titulo,
      'descricao': descricao,
      'tipo_servico': tipoServico,
      'orcamento': orcamento,
      'prazo': prazo,
      'referencia': referencia,
      'status': status,
      'criado_em': criadoEm,
      'atualizado_em': atualizadoEm,
    };
  }
}
